### PR TITLE
Add metrics for parquet page level skipping

### LINKED ISSUE #4086: Add metrics for parquet page level skipping
Usecase is that as we work on https://github.com/apache/arrow-datafusion/issues/4085 we need to have some way to:
1. Verify page index filtering is actually working / enabled as expected
2. Evaluate its efficiency (how long it takes, etc) to see if further optimizations are needed

What I would like is:
1. Add metrics for "how many rows were pruned using page level skipping" and "how long did the page level skipping code take"

Related work is this PR that added similar metrics for filter pushdown (see https://github.com/apache/arrow-datafusion/pull/3989) -- we can probably follow a similar approach 

cc @Ted-Jiang

### PR DESCRIPTION
**Pull request by Ted-Jiang**
Signed-off-by: yangjiang <yangjiang@ebay.com>

# Which issue does this PR close?

Closes #4086 .

# Rationale for this change

we can get metric like
```
ParquetExec: limit=None, partitions=[Users/yangjiang/test-data/1g_tpch_pageIndex/lineitem/part-00000-7d2abab2-a301-4452-9f1d-c641e7f15af4-c000.snappy.parquet],
  predicate=l_shipdate_min@0 <= 8037, projection=[l_quantity, l_extendedprice, l_discount, l_tax, l_returnflag, l_linestatus, l_shipdate], metrics=[output_rows=1020000, elapsed_compute=1ns, spill_count=0, spilled_bytes=0, mem_used=0, 
  pushdown_rows_filtered{filename=Users/yangjiang/test-data/1g_tpch_pageIndex/lineitem/part-00000-7d2abab2-a301-4452-9f1d-c641e7f15af4-c000.snappy.parquet}=0, 
   page_index_rows_filtered{filename=Users/yangjiang/test-data/1g_tpch_pageIndex/lineitem/part-00000-7d2abab2-a301-4452-9f1d-c641e7f15af4-c000.snappy.parquet}=5000000, predicate_evaluation_errors{filename=Users/yangjiang/test-data/1g_tpch_pageIndex/lineitem/part-00000-7d2abab2-a301-4452-9f1d-c641e7f15af4-c000.snappy.parquet}=0, row_groups_pruned{filename=Users/yangjiang/test-data/1g_tpch_pageIndex/lineitem/part-00000-7d2abab2-a301-4452-9f1d-c641e7f15af4-c000.snappy.parquet}=0, bytes_scanned{filename=Users/yangjiang/test-data/1g_tpch_pageIndex/lineitem/part-00000-7d2abab2-a301-4452-9f1d-c641e7f15af4-c000.snappy.parquet}=13198873, num_predicate_creation_errors=0, time_elapsed_scanning=114.532731ms, time_elapsed_processing=3.561757295s, pushdown_eval_time{filename=Users/yangjiang/test-data/1g_tpch_pageIndex/lineitem/part-00000-7d2abab2-a301-4452-9f1d-c641e7f15af4-c000.snappy.parquet}=2ns, 

   page_index_eval_time{filename=Users/yangjiang/test-data/1g_tpch_pageIndex/lineitem/part-00000-7d2abab2-a301-4452-9f1d-c641e7f15af4-c000.snappy.parquet}=475.372µs, time_elapsed_opening=58.666766ms] |
```

# What changes are included in this PR?

# Are there any user-facing changes?

We i try to create single column with multi pages,  i think there is a bug in page size check.  but i found https://github.com/apache/arrow-rs/issues/2941 add  page row number check(not release).

### CODE REVIEW

File: datafusion/core/src/physical_plan/file_format/parquet.rs [hunk 1]
```diff
@@ -1695,6 +1709,35 @@ mod tests {
...
+        let batch2 = create_batch(vec![("int", c2.clone())]);
+
+        let filter = col("int").eq(lit(4_i32));
+
+        let rt =
+            round_trip(vec![batch1, batch2], None, None, Some(filter), false, true).await;
+
+        let metrics = rt.parquet_exec.metrics().unwrap();
+
+        // todo fix this  https://github.com/apache/arrow-rs/issues/2941 release change to row limit.
```
Comment (Ted-Jiang) [inline datafusion/core/src/physical_plan/file_format/parquet.rs:1726]: https://github.com/apache/arrow-rs/pull/2942/files#r1013838557
I think theres is a bug in `should_add_data_page `  `self.encoder.num_values() ` always zero.
So no matter how to set  `data_pagesize_limit ` and `write_batch_size` always return 1 page in on colunn chunk.
🤔 I think someone metion this before.

Comment (Ted-Jiang) [reply datafusion/core/src/physical_plan/file_format/parquet.rs:1726]: But from the real world file, this metic works fine😂

File: datafusion/core/src/physical_plan/file_format/parquet.rs [hunk 2]
```diff
@@ -449,7 +449,7 @@ impl FileOpener for ParquetOpener {
             // page index pruning: if all data on individual pages can
             // be ruled using page metadata, rows from other columns
             // with that range can be skipped as well
-            if let Some(row_selection) = enable_page_index
+            if let Some(row_selection) = (enable_page_index && !row_groups.is_empty())
```
Comment (Ted-Jiang) [inline datafusion/core/src/physical_plan/file_format/parquet.rs:452]: when all rowGroups pruned by rg_metadata, we do this skip fast

**Review by alamb — APPROVED** [review]
Looks good to me @Ted-Jiang  -- thank you 🎉 

Note that https://github.com/apache/arrow-rs/issues/2941 should be included as part of https://github.com/apache/arrow-datafusion/pull/4039 so hopefully we can fix it soon.

I think it would be fine to either merge this PR as is and fix the test as a follow on, or else wait for https://github.com/apache/arrow-datafusion/pull/4039 to merge and then update this PR with the better stats.

**Comment by Ted-Jiang** [comment]
@alamb  fix ut in [use page row limit fix ut](https://github.com/apache/arrow-datafusion/pull/4105/commits/005b10140e0b2da6340d2070c0d6086b475a4f9e)

**Review by alamb — APPROVED** [review]
giphy

**Pull request merged**